### PR TITLE
[GEOS-10434]-Externalized geoserver-environment.properties.

### PR DESCRIPTION
**Pull request by taba90**
[GEOS-10434](https://osgeo-org.atlassian.net/browse/GEOS-10434)

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [x] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/platform/src/main/java/org/geoserver/platform/GeoServerEnvironment.java [hunk 1]
```diff
@@ -97,29 +103,57 @@ public Properties getProps() {
...
-                    Level.WARNING,
-                    "Could not find any '" + PROPERTYFILENAME + "' property file.",
-                    e);
+            String file = propertiesPath != null ? propertiesPath : PROPERTYFILENAME;
+            LOGGER.log(Level.WARNING, "Could not find any '" + file + "' property file.", e);
             props = new Properties();
         }
     }
 
+    private FileWatcher<Properties> getFromPathOrDataDir(String propertiesFilePath) {
```
Comment (aaime) [inline src/platform/src/main/java/org/geoserver/platform/GeoServerEnvironment.java:115]: Name after what a method does, not based on how it does it (we might extend it later to load from a S3 bucket, for what I know).

```suggestion
    private FileWatcher<Properties> getEnvironmentProperties(String propertiesFilePath) {
```

File: src/platform/src/main/java/org/geoserver/platform/GeoServerEnvironment.java [hunk 2]
```diff
@@ -97,29 +103,57 @@ public Properties getProps() {
...
+                result = loadGeoServerEnvProps(PROPERTYFILENAME);
+                String message =
+                        "File "
+                                + propertiesPath
+                                + " not found. Trying to load the environment properties from GeoServer datadir.";
+                LOGGER.warning(message);
+            }
+        } else {
+            result = loadGeoServerEnvProps(PROPERTYFILENAME);
+        }
```
Comment (aaime) [inline src/platform/src/main/java/org/geoserver/platform/GeoServerEnvironment.java:131]:
```suggestion
        if (propertiesFilePath != null && !"".equals(propertiesFilePath)) {
            File propertyFile = new File(propertiesFilePath);
            if (propertyFile.exists()) {
                return loadGeoServerEnvProps(propertyFile);
            } else {
                String message =
                        "File "
                                + propertiesPath
                                + " not found. Trying to load the environment properties from GeoServer datadir.";
                LOGGER.warning(message);
            }
        } 
        return loadGeoServerEnvProps(PROPERTYFILENAME);
```

File: doc/en/user/source/datadirectory/configtemplate.rst [hunk 3]
```diff
@@ -24,7 +24,27 @@ First of all to be able to use this feature, set the following flag via system v
 Then create a file called ``geoserver-environment.properties`` in the root of GeoServer's datadir. 
 This file will contain the definitions for the variables parameterized in the catalog configuration.
 
-Now edit GeoServer's configuration files of the source machine that you want to be parametric, for example let's parameterize the URL of a store 
+We can also place the ``geoserver-environment.properties`` outside the root of GeoServer's datadir. In this case the path to the  ``properties`` file must be defined in one of the following ways:
```
Comment (aaime) [inline doc/en/user/source/datadirectory/configtemplate.rst:27]: The documentation should be written in the third person, not using "we" or "you"... but I see the rest of the document is already written using them... up to you to decide if leave it as is, or fix the whole page.

**Review by aaime — CHANGES_REQUESTED** [review]
Looks good, see couple of comments below.

File: src/platform/src/main/java/org/geoserver/platform/GeoServerEnvironment.java (hunk 1 above)
Comment (taba90) [reply src/platform/src/main/java/org/geoserver/platform/GeoServerEnvironment.java]: fixed

File: doc/en/user/source/datadirectory/configtemplate.rst (hunk 3 above)
Comment (taba90) [reply doc/en/user/source/datadirectory/configtemplate.rst]: fixed

**Comment by taba90** [comment]
thanks for the review @aaime feedback applied.

**Review by aaime — APPROVED** [review]

**Pull request merged**